### ISSUE TITLE
Stuck on PAUSED

### ISSUE THREAD

**Issue by LouDnl** [henrikenblom/SIDPod#9]
Hi there, great project! I have an issue where the device will never leave PAUSED.
Everything seems to work and be connected correctly, display works, rotary works, can set volume choose song etc.
But the double click seems to not do anything

I am using your pre compiled PICO binary.

**Comment by LouDnl** [comment]
Its not the doubleclick I think, it just will not play any sid. Locked specifically for psid files. They show and I can select them, but then its paused immediately

**Comment by LouDnl** [comment]
After compiling myself it unfortunately still did not work.
Tried debugging with gdb but was missing debug symbols. Recompiled with debugsymbols and used the elf with gdb but didnt works, forgot to flash the new compiled version.

As soon as I flashed the compiled bin and started gdb everything worked without doing anything 🤷🏻‍♂️

Still don't know what fixed it 🤣

**Comment by LouDnl** [comment]
![20240301_214452.jpg](https://github.com/henrikenblom/SIDPod/assets/8543038/9e33b72c-fb24-434e-8209-f71524ab2473)

**Comment by MaoCiao7** [comment]
> Just an FYI for future travelers / interested folks: I hit this too.
> 
> TL;DR - make sure that compiler optimizations are turned off in your build, I assume via `cmake -B cmake-build-debug`
> 
> Details:
> 
> (I'm new to Pico dev and using this project as an educational introduction. Along with other unfamiliar tools / tech I'm using VS Code on Windows with various Pico plugins, none of which I really understand at this point so I've been poking things with sticks and using it as an educational exercise, recorded below)
> 
> VS Code just invoked cmake "as is" as is automatically for me, i.e. not as per the instructions provided by [@henrikenblom](https://github.com/henrikenblom). I don't know if `-B cmake-build-debug` would have solved this for me (I assume it probably would), and I didn't know how to make VS code pass the argument, but the result was that the CMakeCache.txt was setup for a full release build with various compiler optimisations left in place.
> 
> One of the compiler optimisations breaks `SIDPlayer.cc`, specifically `SIDPlayer::core1Main()`: Some debugging with Pico Probe and disassembly showed an infinite loop around the `if (playPauseQueued) {` check. It was an unconditional b.n branch to itself. This prevents the play command being acted on. This is presumably because the compiler could see no way that `playPauseQueued` might change so there was no point in running any of the logic after a certain point. As `playPausedQueued` is modified in an IRQ handler, it needs to be marked `volatile` (or similar, I'm not an experienced C++ programmer). Adding the `volatile` to the declaration fixes the play/pause issue for me.
> 
> Investigating another issue led me to changing `build\CMakeCache.txt` manually to use `MAKE_BUILD_TYPE=RelWithDebInfo` and `PICO_DEOPTIMIZED_DEBUG=1`, at which point the problem "went away", i.e. the `volatile` modifier was no longer required (consistent with [@LouDnl](https://github.com/LouDnl)'s comment above and presumably why [@henrikenblom](https://github.com/henrikenblom) wasn't aware of this). So, speaking strictly, the code is not "correct", but it's still wonderful - I'm learning a lot and having a blast.
> 
> FTR: The other issue I've seen which I think must also relate to compiler optimisations is that, with optimisations turned on, after a short period of playing SID audio (approx. 1-2 minutes, but variable), the physical control becomes non-responsive. I've tracked this down in the debugger to timer callbacks no longer firing regularly. It's as if the timer isn't rescheduled when it should be. One of the timers configured `UI.cpp` polls the rotary encoder, so that explains the non-responsiveness. Music continues to play / the screen continues to update because they're not dependent on timers. I _think_ the timers only break when music is playing, not if the playback is paused. Turning optimisations off fixes this issue for me too, i.e. music will play for 10s of minutes with the UI control remaining responsive. I'm currently unclear whether the problem lies in the SDK or SIDPod code. It will take some stubbornness on my part to track this down and I don't know if I have the patience.

Would you please share your forked binary? :) I tried the prebuilt from henrikenblom, but have the same problems.

**Comment by henrikenblom** [comment]
Hey there! First of all, it warms my heart that people are building their own SIDPods! Secondly, I apologize for my lack of activity and engagement. And lastly, I decided to give this project some love, by:

- fixing bugs
- integrating reSID (with tripple-SID support)
- adding SD-card support
- adding playlist support

This all lives in the 'resid' branch (please note that the I2S pins for the display is different in that branch, if you decide to try it out). But since you're here due to bugs, I'll try to find some time in the next few days to fix the most critical ones in the master branch as well. And I'll make sure to keep you posted.

**Comment by henrikenblom** [comment]
Hello again, @MaoCiao7 et al. I've put together a pre-release - version 2.0beta. Give it a try!
https://github.com/henrikenblom/SIDPod/releases

**Comment by MaoCiao7** [comment]
Hi @henrikenblom, 

thank you very much! I've only been able to test a little so far, but it works much better now! Music is coming to the headphones and the navigation is also working. However, only 2 or 3 PSIDs (out of about 10) that I randomly selected for the test also had sound. Maybe I was just unlucky here in terms of compatibility. When there was sound, it was unfortunately not flawless. It seems that not all channels are equally loud and the sounds are generally distorted. Slow CLK issue? For the not working files, the small play animation appears next to the file name in the song selection screen, but otherwise it remains completely silent. In the Visualization Screen, the visual effects are also missing for those files, in line with the non-existent music, but the text is displayed.
I´ll try more songs as soon as possible.

**Comment by henrikenblom** [comment]
> When there was sound, it was unfortunately not flawless. It seems that not all channels are equally loud and the sounds are generally distorted. Slow CLK issue?

This sounds to me like an issue with the amplifier. The hardware setup I've described is a workaround for the lack of I2S headphone amplification breakout boards on the market. - The two resistors between the amp and the headphones are put there to compensate for the difference in impedance specification between headphones and "regular" speakers. But this is basically a very crude hardware hack, that might not work for all headphones. So, could you try to remove the resistors and use a small 2-5W rated speaker instead, to see if the problem persists? If you're using the exact amp model described in the setup, there's also a gain setting to play around with.

> However, only 2 or 3 PSIDs (out of about 10) that I randomly selected for the test also had sound.

Could you post the names, (and artists) of some of the PSIDs that didn't work? So that I can check wether it's due to any of the known issues.

**Comment by MaoCiao7** [comment]
you were right, the distortion was caused by my headphones. I tried my other headset and it sounds clean now. songs with no sound are eg. from Chris Huelsbeck, To be on top & Being boild.

**Comment by henrikenblom** [comment]
My suspicions regarding Ikari Warriors were correct. And I've merged a small fix for it. It turned out to be an unfortunate, and somewhat exotic, combination of a high CIA Timer value and NTSC. So even if it would have played, the tempo would have been completely off.

New release here: https://github.com/henrikenblom/SIDPod/releases/tag/v2.0.1beta

Btw, I'm closing this issue now, since it no longer has anything to do with the original one. But feel free to open new one's at any time.
